### PR TITLE
Implement the rpc service for workers to get AllReduce configuration.

### LINKED ISSUE #2227: New RPC call in master for workers to get allreduce configs (rendezvous port, rank, size, etc).

### PR DESCRIPTION
**Pull request by workingloong**
Fix #2227

### CODE REVIEW

File: elasticdl/proto/elasticdl.proto [hunk 1]
```diff
@@ -94,13 +94,27 @@ message ReportVersionRequest {
...
+}
+
 service Master {
   rpc get_task(GetTaskRequest) returns (Task);
   rpc report_evaluation_metrics(ReportEvaluationMetricsRequest)
       returns (google.protobuf.Empty);
   rpc report_task_result(ReportTaskResultRequest)
       returns (google.protobuf.Empty);
   rpc report_version(ReportVersionRequest) returns (google.protobuf.Empty);
+  rpc get_comm_rank(GetRankRequest) returns (GetRankResponse);
```
Comment (brightcoder01) [inline elasticdl/proto/elasticdl.proto]: get_communication_rank?

File: elasticdl/python/master/master.py [hunk 2]
```diff
@@ -220,6 +222,7 @@ def prepare(self):
             self.instance_manager.update_status(InstanceManagerStatus.PENDING)
             if self.distribution_strategy == DistributionStrategy.ALLREDUCE:
                 # Exposes the consensus service for allreduce-based training
+                self.rendezvous_server.start()
```
Comment (brightcoder01) [inline elasticdl/python/master/master.py:228]: When will we close it?

File: elasticdl/proto/elasticdl.proto (hunk 1 above)
Comment (workingloong) [reply elasticdl/proto/elasticdl.proto]: We can use the abbreviation "comm" to stand for "communicator" like the method [MPI_Comm_rank](https://www.mpich.org/static/docs/v3.2/www3/MPI_Comm_rank.html) in MPI.

File: elasticdl/python/master/master.py (hunk 2 above)
Comment (workingloong) [reply elasticdl/python/master/master.py:228]: We can close it when the job finishes. However, the master will exit if the job finishes, so we don't need to explicitly close it.

File: elasticdl/python/master/master.py [hunk 3]
```diff
@@ -220,6 +222,7 @@ def prepare(self):
...
                 self.instance_manager.start_ftlib_consensus_service()
```
Comment (terrytangyuan) [inline elasticdl/python/master/master.py:229]: Are we starting FTLib service as well in the meantime?

Comment (workingloong) [reply elasticdl/python/master/master.py:229]: We are developing AllReduce using Horovod, so we reserve FTLib service for the moment.

File: elasticdl/python/master/rendezvous_server.py [hunk 4]
```diff
@@ -11,20 +11,29 @@
...
+try:
+    from horovod.runner.common.util.hosts import (
+        get_host_assignments,
+        parse_hosts,
+    )
+    from horovod.runner.http.http_server import RendezvousServer
+
+    _HOROVOD_INSTALLED = True
+except ImportError:
+    _HOROVOD_INSTALLED = False
```
Comment (brightcoder01) [inline elasticdl/python/master/rendezvous_server.py:23]: What behavior do we expect in master and workers if horovod is not installed? Throw exception or switch to fallback strategy?

Comment (workingloong) [reply elasticdl/python/master/rendezvous_server.py:23]: When we use `ParameterServerStrategy`,  we don't need to install Horovod. And the master and worker will not call the module about Horovod like `RendezvousServer` and `training_with_allreduce` if the strategy is `ParameterServerStrategy`.  Later, we need to split PS and AllReduce codes into multiple files not in `worker.py`

Comment (brightcoder01) [reply elasticdl/python/master/rendezvous_server.py:23]: What will happen if we use AllReduceStrategy and HOROVOD is not installed at that time?

File: elasticdl/proto/elasticdl.proto [hunk 5]
```diff
@@ -94,13 +94,25 @@ message ReportVersionRequest {
...
```
Comment (skydoorkai) [inline elasticdl/proto/elasticdl.proto]: `GetCommRankRequest`, `GetCommRankResponse`

File: elasticdl/python/master/rendezvous_server.py (hunk 4 above)
Comment (workingloong) [reply elasticdl/python/master/rendezvous_server.py:23]: The master will raise errors.

File: elasticdl/python/master/servicer.py [hunk 6]
```diff
@@ -145,3 +146,16 @@ def get_average_task_complete_time(self):
...
         return self._worker_liveness_time[worker_id]
+
+    def get_comm_rank(self, request, _):
+        worker_id = request.worker_id
+        k8s_client = self._instance_manager._k8s_client
+        worker_address_port = k8s_client.get_worker_service_address(worker_id)
+        worker_host = worker_address_port.split(":")[0]
+
+        res = elasticdl_pb2.GetRankResponse()
+        res.rank_id = self._rendezvous_server.get_worker_host_rank(worker_host)
```
Comment (skydoorkai) [inline elasticdl/python/master/servicer.py:157]: What will happen if `worker_host` is not in _rendezvous_server's worker host list?
Crash?

Comment (workingloong) [reply elasticdl/python/master/servicer.py:157]: Yes, we can return -1 to tell the worker to retry.

File: elasticdl/python/master/rendezvous_server.py [hunk 7]
```diff
@@ -60,6 +68,10 @@ def get_rendezvous_port(self):
         return self._rendezvous_port
 
     def get_worker_host_rank(self, host):
+        # -1 stands for the worker not in the rendezvous
```
Comment (skydoorkai) [inline elasticdl/python/master/rendezvous_server.py]: -1 if `host` not in worker_hosts list.

**Review by skydoorkai — DISMISSED** [review]

**Review by skydoorkai — APPROVED** [review]

**Pull request merged**